daml-foundations/daml-ghc/ghc-compiler/src/DA/Daml/GHC/Compiler/Upgrade.hs

```diff
@@ -51,32 +69,10 @@ generateGenInstancesModule qual (pkg, L _l src) =
     hasGenDerivation XHsDataDefn{} = False
 
 -- | Generate non-consuming choices to upgrade all templates defined in the module.
-generateUpgradeModule ::
-       (String, ParsedSource) -> (String, ParsedSource) -> String
-generateUpgradeModule (pkgA, L _lA srcA) (pkgB, L _lB srcB) =
-    unlines $ header ++ concatMap upgradeTemplate names
+generateUpgradeModule :: [String] -> String -> String -> String -> String
+generateUpgradeModule templateNames modName pkgA pkgB =
```

Comment:
Rather than generate a Haskell source tree I'd have generated string output directly. I find that's more concise and easy to follow. Of course, more error prone and less types too. Your choice ultimately.

Reply:
So in this particular case we already create a string. For the data declarations, I would prefer  going over the Haskell source tree because we can reuse the generic instances codegen copied from ghc, also the conversion is rather painless and straight forward.